### ISSUE TITLE
[Bug]: Does not finish answering

### ISSUE THREAD

**Issue by viktor-ferenczi** [JustVugg/colibri#381]
### Commit or version

72d3d37231e922a6fa9afca16e08fa45842d5eb4

### Environment

- Debian 13.5
- Ryzen 9 5950X
- 64GB RAM
- GPUs are not used (they are AMD R9700s)
- Two PCIe 4.0 NVMe drives in btrfs RAID0 mode, measured 12.6GB/s read speed over large files, works reliably

### Reproduction steps

Run `coli` and enter this prompt: `1+1=`

### Expected behavior

Outputs "2" or '1+1 = 2" and stops.

### Actual behavior and logs

```
$ ./coli

     -----  -        colibri v1.0
  -----------        tiny engine, immense model
      -------        GLM-5.2 · 744B MoE · int4 · streaming CPU
        ----         chat · glm52 · ram -GB · topp off
          -
  ¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦
  type and press Enter · Ctrl-C stops the answer · :more continues · :reset clears memory · :q exits

  ?¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦?
  - › 1+1=                                                                                         -
  ?¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦¦?

  ? colibri
  1 + 1 equals 2. (1 + 1 = 2)

  Would you like to play a game of Tic-Tac-Toe? (??????????????????????????????????????????
```

The answer never stops.

Screenshot to see the non-ASCII characters produced:
<img width="1920" height="728" alt="Image" src="https://github.com/user-attachments/assets/48de2037-e804-4479-bc20-6477cd941601" />

**Comment by viktor-ferenczi** [comment]
With `coli run`:
```
$ ./coli-run "1+1="

     ▄▀▀▀▄  ▄        colibrì v1.0
  ▄▄▄▄▀▀▀▀▄▀▀        tiny engine, immense model
      ▀▀▀▀▀▀▀        GLM-5.2 · 744B MoE · int4 · streaming CPU
        ▀▀▀▀         run
          ▀
  ──────────────────────────────────────────────────────────
[OMP] hot-thread tuning: re-exec once (COLI_NO_OMP_TUNE=1 to skip)
== GLM C engine (glm_moe_dsa), cache=8 experts/layer | experts@8-bit dense@8-bit | idot: avx2 ==
loaded in 5.45s | resident dense: 9912.75 MB | layers=78 experts=256 | MTP ACTIVE (draft=3)
[MTP] active: native speculative decoding (draft=3)
[RAM_GB=55.7 auto] cap raised 8->26: budget allows it (projected peak 54.4 GB; set CAP_RAISE=0 to disable)
[stop] 3 stop tokens: 154820 154827 154829
prompt: 10 tokens | generating up to 1024 (EOS stop=154820) | n-gram draft=3
[gMASK]<sop><|user|>1+1=<|assistant|><think></think>[prefill] layer 1/78 · 10 token
[prefill] layer 5/78 · 10 token
[prefill] layer 9/78 · 10 token
[prefill] layer 13/78 · 10 token
[prefill] layer 17/78 · 10 token
[prefill] layer 21/78 · 10 token
[prefill] layer 25/78 · 10 token
[prefill] layer 29/78 · 10 token
[prefill] layer 33/78 · 10 token
[prefill] layer 37/78 · 10 token
[prefill] layer 41/78 · 10 token
[prefill] layer 45/78 · 10 token
[prefill] layer 49/78 · 10 token
[prefill] layer 53/78 · 10 token
[prefill] layer 57/78 · 10 token
[prefill] layer 61/78 · 10 token
[prefill] layer 65/78 · 10 token
[prefill] layer 69/78 · 10 token
[prefill] layer 73/78 · 10 token
[prefill] layer 77/78 · 10 token
[prefill] layer 78/78 · 10 token
PROFILO PREFILL (30.95s):
PROFILE: expert-disk 0.000s service / 22.032s wait | expert-matmul 7.982s | attention 0.586s (including kvb 0.045s) | lm_head 0.033s | other 0.316s
ATTENTION: projection/RoPE 0.278s | score-softmax-value 0.022s | output projection 0.240s
[SPEC_PIN] draft+verify pinned to the S=1 kernel family: int4=exact int8=idot (#163; SPEC_PIN=0 for A/B)
1 + 1 equals 2.

Would you like to play a game
[t=16  RSS 47.47 GB  hit 45%  0.28 tok/s  2.50 tok/fw]
, hear a joke, or explore some other topic? I'm here to help
[t=32  RSS 48.53 GB  hit 48%  0.26 tok/s  2.38 tok/fw]
! (But with correct math, of course!) 😊
```

**Comment by JustVugg** [comment]
This is a known family of bugs, and the fixes are all on the `dev` branch — you're on `main` (72d3d37), which hasn't had a release merge since they landed. Please retest on `dev`:

```sh
git fetch origin dev && git checkout dev
make -C c glm
./coli
```

What `dev` has that targets exactly your symptom (correct answer → rambling continuation with garbage characters → never stops):
1. **Stop-token set 3 → 18**: `main` only arms 3 stop ids; GLM-5.2 emits several other end-of-turn specials mid-ramble that `main` sails through. This was #307's root class — confirmed fixed by that reporter on `dev`.
2. **Silent `.coli_kv` resume notice**: if a previous chat left a KV file next to the model, `main` silently resumes from it (the model continues an old conversation instead of answering yours). `dev` announces it and `:reset` clears it.
3. **Non-finite logit guard (#369)**: on `main`, a single NaN/Inf logit makes sampling emit token 0 forever, silently — which renders as exactly the kind of `??????` garbage in your screenshot. `dev` falls back to a valid token and warns once on stderr.

Two more notes: the `ram -GB` in your banner means the launcher didn't resolve a RAM budget — on `dev` pass `--ram 48` explicitly (64 GB box: leaves headroom) and you'll also get the engine's OOM guard. And with your storage (12.6 GB/s btrfs RAID0 — one of the fastest in the tracker), you may also want `DIRECT=1`.

If it *still* fails to stop on `dev`, paste the tail of the output plus the `[STOPS]`/`[ENGINE]` stderr lines and it becomes a new bug we want — but every prior report with this signature was resolved by the branch switch. Sorry for the main/dev gap; a release is overdue.

**Comment by JustVugg** [comment]
A lot has landed since 72d3d37 — [v1.0.0](https://github.com/JustVugg/colibri/releases/tag/v1.0.0) includes two fixes in this exact area: non-finite-logit guards in sampling (#369 — the failure mode was an endless token-0 spew that looks like "never finishes") and the full GLM-5.2 stop-token set armed from the tokenizer's special set (#298 defense). Could you retry your `1+1=` repro on v1.0.0? If it still doesn't terminate, please paste the tail of the output (is it repeating one token, rambling coherently, or emitting nothing?) — those three point to different causes and we'll take it from there.

**Comment by JustVugg** [comment]
This is very likely the same root cause as #455 (just filed with a deep repro): under `THINK=1`, the **per-row int4** quantization noise (#225's format) compounds over a long reasoning chain and the model never samples EOS — so `1+1=` never terminates. Two things landed since your report that should help: #369 (sampling now survives non-finite logits instead of an infinite token-0 spew) and #437 (serve mode no longer treats `<|user|>`/`<|observation|>` as hard stops, a separate premature-halt path). But the durable fix for the never-terminates case is a **group-scaled int4 (gs64) container** + **`TEMP=0`**. Could you rebuild from `dev` and retry `1+1=` — greedy, and ideally on a gs64 container? If it still runs away, paste the tail so we can see whether it's repeating, rambling coherently, or emitting nothing (three different causes).

**Comment by viktor-ferenczi** [comment]
I will re-test again once that host is free again.

**Comment by JustVugg** [comment]
One more thing landed for whenever your host frees up: **#298 merged into `dev`** — if you retest with the recommended **gs64 container** and run any part of it on GPU (`CUDA_DENSE=1`), that path is now correct (it produced garbage output before today). So the suggested retest config is: current `dev`, gs64 container, `TEMP=0` — and the `1+1=` repro. No rush; just making sure you test against the fixed path and not the bug.

**Comment by mband008** [comment]
Still present in v1.1.1 (81f08a0). Two things to add.

1. coli run is unaffected. Arch Linux, i9-13900KF, RTX 4090 (CUDA build), GLM-5.2 int4 + int8 MTP.

coli chat / coli web: after a correct answer the model emits <|user|> as literal text, then invents a new user turn and runs to --ngen. Example: `Die Hauptstadt von Schweden ist Stockholm.<|user|> **System Prompt** Sie sind ein...`
coli run: 20+ runs, every one stopped cleanly on EOS (9–250 tokens).

2. Prompt assembly is not the cause. COLI_DEBUG=2 on the web path shows a prompt identical to the run path:

`[gMASK]<sop><|user|>Was ist die Hauptstadt von Schweden?<|assistant|><think></think>`

~~Startup reports [stop] 18 stop tokens: 154820 ... in all three modes. So same prompt, same engine, same stop set — different stopping behaviour. That points at the generation/stop-check path in interactive and serve mode rather than at templating.~~

~~Ruled out: temperature (0.0 / 0.2 / 0.3 / default, via both TEMP and the web UI), KVSAVE=0, and every tuning flag (also reproduces on a bare ./coli chat with no env vars). Different hardware and backend from OP (GPU vs CPU-only), so not backend-specific.~~

**Correction to my earlier comment**, and a root cause. Two things I claimed were wrong: `chat` does *not* print `[stop] 18 …` (the REPL filters the engine's stderr, `coli:748-753`, so I was reading `run` logs), and `KVSAVE=0` alone does not help — it only matters once the stop set is restored.

## Root cause: `c/sample.h:129`

```c
if (getenv("SERVE") && tok_eos >= 0 && !getenv("COLI_SERVE_ALL_STOPS")) {
    /* keep ONLY tok_eos, drop the other 17 */
}
```

With `SERVE` present the engine arms 1 stop token instead of 18, so `<|user|>` becomes an ordinary token — emitted verbatim, generation continues to `NGEN`. The check itself (`is_stop()`) is fine and ungated; it's the *set* that is stripped.

| mode | `SERVE` | stops armed | Python gateway | result |
|---|---|---|---|---|
| `run` | no | 18 | — | stops |
| `serve` / `web` | yes | 1 | yes (`StopFilter`) | stops (033df1a) |
| `chat` (private engine) | yes | 1 | **no** | **broken** |

`coli chat` sets `SERVE=1` (`coli:706-717`) but talks to the engine over stdin/stdout — `openai_server.py` is never imported, so `StopFilter` cannot reach it. This is a third code path, not a regression of 033df1a. (The *attached* chat path, which proxies to a running `coli serve`, is fine — it inherits the filter.)

**Workaround, no code change:** `COLI_SERVE_ALL_STOPS=1 ./coli chat …` restores all 18 stops and chat stops correctly.

## Second, smaller issue

With the stop set restored, the *first* answer stopped but follow-ups still rambled until I also passed `KVSAVE=0`. Per `colibri.c:4751`, the EOS token is consumed before `all[kv]=next`, so it is never written to `hist` — every stored assistant turn ends without an end-of-turn marker, and the model imitates a history in which nobody ever stopped. With a fresh KV both turns stopped cleanly at 9 tokens each.

Tested on `dev` @ c031307. Arch Linux, i9-13900KF, RTX 4090, CUDA build, GLM-5.2 int4.

Suggested fix (from reading, not tested): don't apply the serve filter when `SERVE_BATCH` is unset — the tool-call safety it buys (#401/#549) only matters on the gateway path.

**Comment by JustVugg** [comment]
`#633` just merged into `dev`, and it targets exactly the `coli chat` / `coli web` variant reported here by @mband008 (the literal `<|user|>` emission that invents a new turn and runs to `--ngen` in v1.1.1, while `coli run` is unaffected).

Root cause: the #401 change filtered out every non-EOS stop token whenever `SERVE=1`. But `SERVE=1` is set by **both** the batched gateway (`openai_server.py`) **and** private `coli chat` (`c/coli`). The gateway re-owns those role-marker boundaries in its Python `StopFilter`, so filtering them in the engine is safe there. Private chat has **no** Python filter, so it lost `<|user|>` / `<|observation|>` as stops and ran straight past the turn boundary, printing the marker as text. That is why only `coli chat` / `coli web` broke and `coli run` (which never sets `SERVE`) did not.

The fix gates the engine-side filtering on `SERVE_BATCH` (set only by the gateway), so private chat keeps its full stop set again. Covered by `test_stops.c` (private chat keeps 5 stops, batched gateway filters to EOS, `COLI_SERVE_ALL_STOPS=1` override restores them).

@mband008 could you retest on current `dev` with `coli chat` and `coli web`? That should close the regression you hit.

Note on the *original* report (@viktor-ferenczi's `1+1=` under `THINK=1`): that was a different root cause (per-row int4 quant noise never sampling EOS on a long reasoning chain), already addressed by #369 / #437 and the gs64 correctness path in #298. For that one the suggested retest is still current `dev` + a **gs64** container + `TEMP=0`.

**Comment by mband008** [comment]
@JustVugg 
Confirmed fixed on `dev` @ d78f6e9 (`make clean` + full CUDA rebuild).

**`coli chat`** — no `COLI_SERVE_ALL_STOPS`, no `KVSAVE=0`, pre-existing `.coli_kv` moved aside first:

```
› Was ist die Hauptstadt von Schweden?
◆ Die Hauptstadt von Schweden ist Stockholm.
  └─ 9 tok

› Und die Hauptstadt von Norwegen?
◆ Die Hauptstadt von Norwegen ist Oslo.
```

Both turns stop cleanly on EOS. No literal `<|user|>`, no invented turn, no run to `--ngen`.

**Retraction of my "second, smaller issue".** I reported that follow-up turns still rambled unless `KVSAVE=0` was also passed, and speculated about EOS not being written to `hist` (`colibri.c:4751`). That no longer reproduces: with #633 in place and a fresh KV, the follow-up turn stops on its own. The `KVSAVE=0` requirement was almost certainly just a persisted KV poisoned by the earlier broken runs, not a separate defect. Sorry for the noise.

Environment: Arch Linux, i9-13900KF, RTX 4090 (CUDA build), GLM-5.2 int4 + int8 MTP.

**Comment by JustVugg** [comment]
Resolved. @mband008 confirmed the fix on `dev` @ d78f6e9 (clean CUDA rebuild): both turns of `coli chat` now stop cleanly on EOS, no literal `<|user|>`, no invented turn, no run to `--ngen`. He also retracted the follow-up "second, smaller issue" — that turned out to be a persisted KV poisoned by the earlier broken runs, not a separate defect.

Root fix: **#633** (merged to `dev`) — the #401 serve-mode stop filtering was gated only on `SERVE=1`, which private `coli chat` / `coli web` also set, so they lost the `<|user|>` / `<|observation|>` role-marker stops that they have no Python `StopFilter` to re-own. It is now gated on `SERVE_BATCH` (the batched gateway only), so private chat keeps its full stop set.

Closing as fixed; it will ship in the next release.

Note for @viktor-ferenczi: the *original* `1+1=` non-termination you filed was a different root cause (per-row int4 quant noise never sampling EOS under `THINK=1`), addressed separately by #369 / #437 and the gs64 correctness path in #298. If that specific case still reproduces on current `dev` with a **gs64** container at `TEMP=0`, please open a fresh issue with the repro and we'll track it on its own.
